test/MemeTest.sol

```diff
@@ -28,41 +36,105 @@ contract MemeTest is TestBaseContract {
     //
     // ------------------------------------------------------ //
 
-    function testBuyMemeWithInsufficientWMatic() public {
+    function testBuyMemeWithInsufficientCoin() public {
         proxy.setAddress(LibConstants.TREASURY_ADDRESS, address(proxy));
 
-        IERC20 wmatic = IERC20(LibConstants.WMATIC_POLYGON_ADDRESS);
-
         vm.expectRevert();
         proxy.buyMeme(MemeBuySizeDollars.Five);
     }
 
-    function testPurchaseMeme() public {
+    function testZkSyncMockPurchaseMeme() public {
+        // deploy mock twap
+        MockTwap twap = new MockTwap();
+
+        twap.setSqrtPriceX96(1625 * 10 ** 18);
+        proxy.setPriceOracle(address(twap));
+
         proxy.setAddress(LibConstants.TREASURY_ADDRESS, address(proxy));
-        writeTokenBalance(
-            address(this),
-            address(proxy),
-            LibConstants.WMATIC_POLYGON_ADDRESS,
-            10000e18
+
+        console2.log(address(this).balance);
+
+        address user1 = address(0x1);
+
+        uint256 user1_startingBalance = user1.balance;
+
+        vm.prank(user1);
+        proxy.buyMeme{ value: 1 ether }(MemeBuySizeDollars.Five);
+        uint256 user1_endingBalance = user1.balance;
+
+        uint256 receiverBalance = address(SERVER_ADDRESS).balance;
+
+        console2.log("receiver balance", receiverBalance);
+        console2.log("user1 ending balance", user1_endingBalance);
+        assertEq(
+            user1_startingBalance,
+            user1_endingBalance + receiverBalance,
+            "user1 balance is not correct"
         );
 
-        IERC20 wmatic = IERC20(LibConstants.WMATIC_POLYGON_ADDRESS);
+        assertEq(memeToken.balanceOf(user1), 1000 ether);
+    }
+
+    function testPurchaseMemeRoutingEthToTreasury() public {
+        // deploy mock twap
+        MockTwap twap = new MockTwap();
 
-        uint256 prevBal = wmatic.balanceOf(address(this));
-        proxy.buyMeme(MemeBuySizeDollars.Five);
+        twap.setSqrtPriceX96(1625 * 10 ** 18);
+        proxy.setPriceOracle(address(twap));
+
+        proxy.setAddress(LibConstants.TREASURY_ADDRESS, address(proxy));
 
-        uint256 newBal = wmatic.balanceOf(address(this));
-        console2.log("new balance", newBal);
-        assertEq(memeToken.balanceOf(address(this)), 1000 ether);
+        console2.log(address(this).balance);
+
+        address user1 = address(0x1);
+
+        // give more than 0.5 coin to server
+        vm.deal(SERVER_ADDRESS, 1 ether);
```

Comment:
Please add a test which does the opposite and check that server gets the funds. Then check that once server balance is over 0.5 ETH it no longer gets funds. Also, let's make the 0.5 ETH a constant in LibConstants so that we don't hardcode it anywhere.

Reply:
have a test for that - testPurchaseMemeRoutingEthToTreasury, but will make it more robust